ctlearn/run_model.py

```diff
@@ -254,23 +252,25 @@ def run_model(config, mode="train", debug=False, log_to_file=False):
     run_tfdbg = config.get('TensorFlow', {}).get('run_TFDBG', False)
 
     # Define input function for TF Estimator
-    def input_fn(reader, indices, output_names, output_dtypes, output_shapes,
-                 label_names, seed=None, batch_size=1,
-                 shuffle_buffer_size=None, prefetch_buffer_size=1,
+    def input_fn(reader, indices, output_names, output_dtypes,
+                 label_names, shuffle_and_repeat=False, seed=None,
+                 batch_size=1, prefetch_to_device=None,
                  add_labels_to_features=False):
 
-        def generator(indices):
-            for idx in indices:
-                yield tuple(reader[idx])
+        dataset = tf.data.Dataset.from_tensor_slices(indices)
+        if shuffle_and_repeat:
+            dataset = dataset.shuffle(buffer_size=len(indices), seed=seed,
+                                      reshuffle_each_iteration=True)
+            dataset = dataset.repeat()
+        dataset = dataset.map(lambda x: tf.py_function(func=reader.__getitem__,
```

Comment:
This is very nice, because we can add a function in the reader to return the output in the shape `(features, labels)`. That would be needed for the TF2 update. Is this true?

Reply:
The output is already returned in this format (see line 288), unless I missed something?

Reply:
Sorry my comment wasn't very precise! With `(features, labels)` I meant a 2D tuple of tuples. The current TF2 input, which is returning a dataset looks like this:
```
def input_fn(reader, indices, output_dtypes, output_shapes,
             feature_positions, label_positions, mode='train', seed=None, batch_size=1,
             shuffle_buffer_size=None, prefetch_buffer_size=1,
             add_labels_to_features=False):
    
    def generator(indices):
        for idx in indices:
            yield (map(lambda fpos: reader[idx][fpos], feature_positions), map(lambda lpos: reader[idx][lpos], label_positions))
    
    dataset = tf.data.Dataset.from_generator(generator, output_dtypes,
                                             output_shapes=output_shapes,
                                             args=(indices,))
   
    # Only shuffle the data, when train mode is selected.
    if mode == 'train':
        if shuffle_buffer_size is None:
            shuffle_buffer_size = len(indices)
        dataset = dataset.shuffle(buffer_size=shuffle_buffer_size, seed=seed)
    dataset = dataset.batch(batch_size)
    dataset = dataset.prefetch(prefetch_buffer_size)
    
    return dataset
```
 What we can do is to add a function`reader.__getitem_TF2__` with arguments `feature_positions` and `label_positions` and splitting the reader output in the right format. Then we can return the the dataset from the `input_fn()` and skip the custom separation from line 275 - 286.